test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -444,6 +484,495 @@ public void testDurableDataLogFail() throws Exception {
+
+        // Create two streams for writing data onto two different segments
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM1);
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM2);
+        log.info("Created two streams.");
+
+        @Cleanup
+        ConnectionFactory connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder()
+                .controllerURI(controllerStarter.controllerURI).build());
+        @Cleanup
+        ClientFactoryImpl clientFactory = new ClientFactoryImpl(SCOPE, controllerStarter.controller, connectionFactory);
+        @Cleanup
+        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl(SCOPE, controllerStarter.controller, clientFactory);
+
+        log.info("Writing events on to stream: {}", STREAM1);
+        writeTransactionalEvents(STREAM1, clientFactory); // write 300 events on one segment
+        log.info("Writing events on to stream: {}", STREAM2);
+        writeTransactionalEvents(STREAM2, clientFactory); // write 300 events on other segment
+
+        // Create two readers for reading both the streams.
+        EventStreamReader<String> reader1 = createReader(clientFactory, readerGroupManager, SCOPE, STREAM1, testReaderGroup1, testReader1);
+        EventStreamReader<String> reader2 = createReader(clientFactory, readerGroupManager, SCOPE, STREAM2, testReaderGroup2, testReader2);
+
+        // Let readers read N number of events and mark their positions.
+        Position p1 = readNEvents(reader1, NUM_EVENTS);
+        Position p2 = readNEvents(reader2, NUM_EVENTS);
+
+        ReaderGroup readerGroup1 = readerGroupManager.getReaderGroup(testReaderGroup1);
+        ReaderGroup readerGroup2 = readerGroupManager.getReaderGroup(testReaderGroup2);
+
+        readerGroup1.readerOffline(testReader1, p1);
+        readerGroup2.readerOffline(testReader2, p2);
+
+        readerGroupManager.close();
+        clientFactory.close();
+
+        controllerStarter.close(); // Shut down the controller
+
+        // Get names of all the segments created.
+        ConcurrentHashMap<String, Boolean> allSegments = segmentStoreStarter.segmentsTracker.getSegments();
+        log.info("No. of segments created = {}", allSegments.size());
+
+        // Get the long term storage from the running pravega instance
+        @Cleanup
+        Storage storage = new AsyncStorageWrapper(new RollingStorage(this.storageFactory.createSyncStorage(),
+                new SegmentRollingPolicy(DEFAULT_ROLLING_SIZE)), executorService());
+
+        // wait for all segments to be flushed to the long term storage.
+        waitForSegmentsInStorage(allSegments.keySet(), segmentStoreStarter.segmentsTracker, storage)
+                .get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+
+        segmentStoreStarter.close(); // Shutdown SegmentStore
+        log.info("Segment Store Shutdown");
+
+        this.bookKeeperStarter.close(); // Shutdown BookKeeper & ZooKeeper
```

Comment:
You really shouldn't be touching the class-level fields in a test. In unit tests, the only places where you can assign, un-assign or otherwise "close" such an object instance is in the setup/teardown methods.

Reply:
I need to reproduce the scenario of bookKeeper shut down, and then starting a new instance.

Reply:
Then you need to create them locally and destroy them in the same scope where you created them. It is never a good practice to repurpose a higher-scoped variable (i.e., class-level field) in this case since your method will have unintended side effects.

Reply:
Thanks. Made changes accordingly.